### PR TITLE
common/PriorityCache: Implement a Cache Manager

### PR DESCRIPTION
**Pull request by markhpc**
Signed-off-by: Mark Nelson <mnelson@redhat.com>

Move cache management out of bluestore and into the PriorityCache namespace

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/common/PriorityCache.h [hunk 1]
```diff
@@ -68,6 +101,42 @@ namespace PriorityCache {
     // Get the name of this cache.
     virtual std::string get_cache_name() const = 0;
   };
+
+  struct Manager {
```
Comment (ifed01) [inline src/common/PriorityCache.h]: Make it a class and hide the internals?

File: src/common/PriorityCache.h [hunk 2]
```diff
@@ -68,6 +101,42 @@ namespace PriorityCache {
...
+    CephContext* cct = nullptr;
+    PerfCounters* logger;
```
Comment (batrick) [inline src/common/PriorityCache.h]: nit: std::unique_ptr<> would be better

File: src/common/PriorityCache.h [hunk 3]
```diff
@@ -68,6 +101,42 @@ namespace PriorityCache {
...
+    std::unordered_map<std::string, PerfCounters*> loggers;
```
Comment (batrick) [inline src/common/PriorityCache.h]: suggest transparent comparator for these maps so you can index with `std::string_view`. i.e.

`std::unordered_map<std::string, PerfCounters*, std::less<>> loggers;`

File: src/common/PriorityCache.h [hunk 4]
```diff
@@ -68,6 +101,42 @@ namespace PriorityCache {
...
+    void set_max_memory(uint64_t max) {
+      max_mem = max;
+    }
+    void set_target_memory(uint64_t target) {
+      target_mem = target;
+    }
+    uint64_t get_tuned_mem() const {
+      return tuned_mem;
+    }
+    void insert(std::string name, std::shared_ptr<PriCache> c);
```
Comment (batrick) [inline src/common/PriorityCache.h]: `std::string_view` instead, and `const std::shared_ptr<PriCache>& c`

File: src/common/PriorityCache.h [hunk 5]
```diff
@@ -68,6 +101,42 @@ namespace PriorityCache {
...
+    void erase(std::string name);
```
Comment (batrick) [inline src/common/PriorityCache.h]: `std::string_view`

File: src/os/bluestore/BlueStore.h [hunk 6]
```diff
@@ -2022,7 +2022,8 @@ class BlueStore : public ObjectStore,
     ceph::mutex lock = ceph::make_mutex("BlueStore::MempoolThread::lock");
     bool stop = false;
     uint64_t autotune_cache_size = 0;
-    std::shared_ptr<PriorityCache::PriCache> binned_kv_cache = nullptr;
+    std::shared_ptr<PriorityCache::PriCache> binned_kv_cache = nullptr;
```
Comment (batrick) [inline src/os/bluestore/BlueStore.h]: churn caused by unnecessary whitespace at EOL

**Comment by sseshasa** [comment]
> Some minor code hygiene changes suggested.
> 
> Other than that, I haven't yet wrapped my head around how to use this code to setup a priority cache for the MDS. At first glance, We should discuss this offline Mark. A walkthrough would be helpful.

I second this. A walk-through would be helpful to set this up for MON.

**Comment by markhpc** [comment]
Hi Guys,

Thanks for the feedback!  I'll review the requested changes.  Would be happy to schedule a meeting to do an overview as well.

In the mean time, the basic idea is that when you've implemented the PriorityCache interface and register your cache with the manager, it can request and compete for memory at different priority levels with other registered priority caches.  The amount of memory that is requested at each priority level is up to each cache.  The amount of memory that is assigned at each priority level is based on a multi-round competition.  In the first round caches only receive up to their "fair share" based on initial percentile targets.  If a cache has it's request fulfilled it is removed from the competition.  If memory is left, any caches that still want memory at that priority level will compete in further rounds until either all requests are satisfied or no memory is left.  The process is repeated at each priority level.  At the end, any remaining memory is assigned based on the initial percentile targets.

To see how this works in the OSD:

// create the cache manager and give it initial min and max (aggregate) cache sizes and the overall process memory target:

pcm = std::make_shared<PriorityCache::Manager>(store->cct, min, max, target);

// Register some caches:

pcm->insert("kv", binned_kv_cache);
pcm->insert("meta", meta_cache);
pcm->insert("data", data_cache);

// Tune the amount of memory available for caches based on process mapped memory
pcm->tune_memory();

// re-balance the caches based on the above described process and current available memory for caches.
pcm->rebalance();

This is sort of the backbone to be able to compare the relative age of cache items across multiple caches using age-binning.  That work will come in another PR, though the code is available here:

https://github.com/markhpc/ceph/commit/2b370fec0a4e4a0ad62506862912f81d09457aa8

Mark

File: src/common/PriorityCache.cc [hunk 7]
```diff
@@ -50,6 +56,313 @@ namespace PriorityCache {
...
+  }
+
+  Manager::~Manager()
+  {
+    std::vector<std::string> vec;
+    for (auto && i : caches) {
+      vec.emplace_back(i.first);
+    }
+    for (auto&& name : vec) {
+      erase(name);
```
[the remark this answers is not in the document]
Comment (markhpc) [reply src/common/PriorityCache.cc]: Ended up adding a clear function that just sets the next iterator in a loop using erase for the logger map and clears the others directly.

File: src/common/PriorityCache.cc [hunk 8]
```diff
@@ -50,6 +56,313 @@ namespace PriorityCache {
...
+    }
+
+    auto l = b.create_perf_counters();
+    loggers.emplace(name, l);
+    cct->get_perfcounters_collection()->add(l);
+
+    cur_index = end;
+  }
+
+  void Manager::erase(std::string name) {
```
[the remark this answers is not in the document]
Comment (markhpc) [reply src/common/PriorityCache.cc]: Yep!

File: src/common/PriorityCache.h (hunk 1 above)
Comment (markhpc) [reply src/common/PriorityCache.h]: sure thing!

File: src/common/PriorityCache.h (hunk 2 above)
Comment (markhpc) [reply src/common/PriorityCache.h]: Going to leave this alone for now.  We use raw pointers for this everywhere else in the code so if we want to make a change let's rethink how logger works generally.

File: src/common/PriorityCache.h (hunk 3 above)
Comment (markhpc) [reply src/common/PriorityCache.h]: See: https://stackoverflow.com/questions/51749081/stdstring-view-and-stdstring-in-stdunordered-set

**Comment by markhpc** [comment]
Noticed a couple of additional bugs as I was testing the changes related to the new reserved memory and 10-level priority handling (This PR originally was part of a multi-commit set that included the new age-binning code).  I think I've got them resolved at this point with the cache balancing looking pretty close to the current behavior in master.

**Comment by neha-ojha** [comment]
http://pulpito.ceph.com/nojha-2019-04-12_14:38:37-rados-wip-27381-distro-basic-smithi/

**Comment by markhpc** [comment]
http://pulpito.ceph.com/nojha-2019-04-12_14:38:37-rados-wip-27381-distro-basic-smithi/3838364/

Am I reading the log correctly that it was ceph_test_rados itself that crashed?

**Comment by liewegas** [comment]
The osd threw an EIO, but it' snot related to this.  I'll open a ticket.

**Comment by neha-ojha** [comment]
> http://pulpito.ceph.com/nojha-2019-04-12_14:38:37-rados-wip-27381-distro-basic-smithi/

None of the failures look related.
